text/0000-derive-default-enum-with-fields.md

```diff
@@ -0,0 +1,272 @@
+However, the degree to which complexity is increased is not substantial. If
+anything, the complexity of the concepts needed to be understood is reduced, as
+there are fewer special cases users need to keep in mind when using
+`#[derive(Default)]`, as well as allow us to remove `impl Default`s from the
+standard library.
+
+[The same](https://github.com/rust-lang/rust/issues/26925) issue highlighted on
+[RFC-3107] of current `#[derive(Default)]` on `struct`s producing `impl`s with
+incorrect bounds (non-[perfect derives]) applies to this proposal as well.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+[`derivative`]: https://crates.io/crates/derivative
+
+[`smart-default`]: https://crates.io/crates/smart-default
+
+As shown by the existence of [`derivative`] and [`smart-default`], there is a
+desire to fill this perceived gap in flexibility that the built-in
+`#[derive(Default)]` support has. We can do nothing and let the ecosystem sort
+this gap out.
+
+
+# Prior art
+[prior-art]: #prior-art
+
+## Procedural macros
+
+There are a number of crates which to varying degrees afford macros for default field values and
+associated facilities.
+
+
+### `#[derive(Derivative)]`
+
+[`derivative`]: https://crates.io/crates/derivative
+
+The crate [`derivative`] provides the `#[derivative(Default)]` attribute. With it, you may write:
+
+```rust
+#[derive(Derivative)]
+#[derivative(Default)]
+enum Foo {
+    #[derivative(Default)]
+    Bar {
+        value: Option<i32>,
+    },
+    Baz,
+}
+```
+
+Contrast this with the equivalent in the style of this RFC:
+
+```rust
+#[derive(Default)]
+enum Foo {
+    #[default]
+    Bar {
+        value: Option<i32>,
+    },
+    Baz,
+}
+```
+
+Like in this RFC, `derivative` allows you to derive `Default` for `enum`s. The
+syntax used in the macro is `#[derivative(Default)]` whereas the RFC provides
+uses the already existing `#[default]` annotation.
+
+### `#[derive(SmartDefault)]`
+
+[`smart-default`]: https://crates.io/crates/smart-default
+
+The [`smart-default`] provides `#[derive(SmartDefault)]` custom derive macro. It functions similarly
+to `derivative` but is specialized for the `Default` trait. With it, you can write:
+
+```rust
+#[derive(SmartDefault)]
+enum Foo {
+    #[default]
+    Bar {
+        value: Option<i32>,
+    },
+    Baz,
+}
+```
+
+- There is no trait `SmartDefault` even though it is being derived. This works because
+  `#[proc_macro_derive(SmartDefault)]` is in fact not tied to any trait. That `#[derive(Serialize)]`
+  refers to the same trait as the name of the macro is from the perspective of the language's static
+  semantics entirely coincidental.
+
+  However, for users who aren't aware of this, it may seem strange that `SmartDefault` should derive
+  for the `Default` trait.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Should we wait until [perfect derives] are addressed first?
```

Comment:
> But again, I don't think we need a bound on the generics used by the default variant, what we would need is a bound on the exact field types used in the variant. The way that the generic is wrapped in other types is important. `i32` is `Default`, but `&i32` is not, so if a field is `&T` then a bound on `T: Default` is the wrong bound.

@Lokathor but it is *the same* wrong bound we have for `struct`s. What you want is perfect derives, which is explicitly called out in the text: do we wait for perfect derives or land support for the same imperfect behavior we have today for `struct`s.

The reason we *can't* do `Option<T>: Default` is because we only have AST information on these macros, and the only way to provide those bounds *correctly* is to detect cycles, both direct (we can detect those) *and* indirect (we can't detect those unless we have resolve/typeck info).

As I'm writing this though, I'm thinking that it *could* be that we could change the desugaring of `derive` to use the more accurate `where` clauses, and have a late lint check for cycles after the fact and tell people to instead write an explicit `impl` instead 🤔 

But that would be its own RFC for "get closer to perfect derives without trait solver changes", and I'm gonna be off-line for several weeks, so I can't drive that now.

Reply:
> do we wait for perfect derives or land support for the same imperfect behavior we have today for structs.

If we can have it be better than before, I don't see a reason to not make it better than before. If it means that `enum` can (temporarily) derive more precisely than `struct` that seems completely fine to me.

I mean we could start with a bound on all the field types *and also* on the generics (that's what structs do now), but we literally can't have a bound *just* on the generics, without the bound on the field types. That actually won't work, which is my point.

Reply:
I'm pretty confused, what do you mean when you say we can't do it / it doesn't work? It is what we have for structs now, why couldn't we do the same thing here? Yes, it's not going to always generate an implementation that passes typecheck. But the same is true for structs today, and AFAIK this is not considered a major problem.

Reply:
That is basically what i mean, yes. That the code will have a build error.